File: figures/src/r/classes/sample_pca.R
repo_root <- getOption("manuscript.figure.repo_root")
if (is.null(repo_root)) {
  stop("manuscript.figure.repo_root option is required")
}
source(file.path(repo_root, "figures", "src", "r", "common.R"))

create_plot <- function(spec_path) {
  spec <- load_yaml(spec_path)
  theme <- load_yaml(file.path(repo_root, "figures/config/project_theme.yml"))
  font_policy <- load_yaml(file.path(repo_root, "figures/config/font_policy.yml"))
  profiles <- load_yaml(file.path(repo_root, "figures/config/venue_profiles.yml"))
  registry <- load_yaml(file.path(repo_root, "figures/registry/classes.yml"))$classes
  profile <- profiles$profiles[[spec$target_profile]]
  class_entry <- registry[[spec$class_id]]
  validate_common_contract(spec, theme, font_policy, profile, class_entry)
  resolved_font <- resolve_font(font_policy)
  plot_theme <- build_theme(theme, resolved_font)

  rows <- read.csv(file.path(repo_root, spec$data_inputs[[1]]), stringsAsFactors = FALSE, check.names = FALSE)
  write_source_csv(
    file.path(repo_root, source_data_mapping(spec)$a),
    rows,
    c("sample_id", "condition", "batch", "pc1", "pc2", "highlight_label")
  )

  plot <- ggplot2::ggplot(rows, ggplot2::aes(x = pc1, y = pc2, color = condition, shape = batch)) +
    ggplot2::geom_hline(yintercept = 0, linewidth = 0.4, color = theme$palette$neutral[[2]]) +
    ggplot2::geom_vline(xintercept = 0, linewidth = 0.4, color = theme$palette$neutral[[2]]) +
    ggplot2::geom_point(size = 2.5, stroke = 0.5) +
    ggplot2::geom_text(
      data = rows[rows$highlight_label == "yes", ],
      ggplot2::aes(x = pc1 + 0.12, y = pc2 + 0.12, label = sample_id),
      inherit.aes = FALSE,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[1]]
    ) +
    ggplot2::scale_color_manual(values = c("Control" = theme$palette$categorical[[1]], "Treatment" = theme$palette$categorical[[2]])) +
    ggplot2::labs(title = "Sample PCA", x = "PC1 (48%)", y = "PC2 (23%)") +
    plot_theme +
    ggplot2::theme(
      plot.tag = ggplot2::element_text(size = as.numeric(theme$panel_labels$font_size_pt), face = theme$panel_labels$font_weight, family = resolved_font$family)
    )
  plot <- plot + patchwork::plot_annotation(tag_levels = "a")
  list(
    plot = plot,
    spec = spec,
    theme = theme,
    profile = profile,
    resolved_font = resolved_font,
    rows = rows
  )
}

build_figure <- function(spec_path) {
  payload <- create_plot(spec_path)
  spec <- payload$spec
  theme <- payload$theme
  profile <- payload$profile
  resolved_font <- payload$resolved_font
  plot <- payload$plot
  rows <- payload$rows
  output_dir <- file.path(repo_root, spec$renderers$r$output_dir)
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  stem <- spec$figure_id
  svg_path <- file.path(output_dir, sprintf("%s.svg", stem))
  pdf_path <- file.path(output_dir, sprintf("%s.pdf", stem))
  png_path <- file.path(output_dir, sprintf("%s.png", stem))
  width_in <- mm_to_inches(as.numeric(spec$size$width_mm))
  height_in <- mm_to_inches(as.numeric(spec$size$height_mm))

  svglite::svglite(file = svg_path, width = width_in, height = height_in, bg = theme$export_defaults$background, system_fonts = list(sans = resolved_font$family))
  print(plot)
  grDevices::dev.off()
  grDevices::cairo_pdf(file = pdf_path, width = width_in, height = height_in, bg = theme$export_defaults$background, family = resolved_font$family)
  print(plot)
  grDevices::dev.off()
  ragg::agg_png(filename = png_path, width = width_in, height = height_in, units = "in", res = as.numeric(profile$preview_dpi), background = theme$export_defaults$background)
  print(plot)
  grDevices::dev.off()

  outputs <- list(svg = sub(paste0("^", repo_root, "/"), "", svg_path), pdf = sub(paste0("^", repo_root, "/"), "", pdf_path), png = sub(paste0("^", repo_root, "/"), "", png_path))
  write_manifest(
    repo_root,
    spec,
    profile,
    resolved_font,
    spec_path,
    "r",
    outputs,
    list(
      "group_coloring_for_sample_separation",
      "batch_shape_encoding",
      "selective_sample_labels",
      "vector_first_export"
    ),
    list(
      panel_count = length(spec$panels),
      highlight_label_count = sum(rows$highlight_label == "yes"),
      reference_line_count = 2
    )
  )
  outputs
}

File: figures/src/python/classes/sample_pca.py
#!/usr/bin/env python3
"""Class-based Python renderer for sample-level PCA plots."""

from __future__ import annotations

import csv
from pathlib import Path
from typing import Any

from matplotlib.figure import Figure

from figures.src.python.common import (
    REPO_ROOT,
    apply_publication_layout,
    configure_matplotlib,
    export_csv,
    mm_to_inches,
    project_resources,
    resolve_font_stack,
    source_data_mapping,
    validate_common_contract,
    write_manifest,
)


def _load_rows(path: Path) -> list[dict[str, Any]]:
    with path.open("r", encoding="utf-8", newline="") as handle:
        reader = csv.DictReader(handle)
        return [
            {
                "sample_id": row["sample_id"],
                "condition": row["condition"],
                "batch": row["batch"],
                "pc1": float(row["pc1"]),
                "pc2": float(row["pc2"]),
                "highlight_label": row["highlight_label"],
            }
            for row in reader
        ]


def build_source_data(spec: dict[str, Any], rows: list[dict[str, Any]]) -> list[dict[str, Any]]:
    export_csv(
        REPO_ROOT / source_data_mapping(spec)["a"],
        rows,
        ["sample_id", "condition", "batch", "pc1", "pc2", "highlight_label"],
    )
    return rows


def create_figure(spec_path: Path) -> Figure:
    spec, theme, font_policy, profile, class_entry = project_resources(spec_path)
    validate_common_contract(spec, theme, font_policy, profile, class_entry)
    rows = build_source_data(spec, _load_rows(REPO_ROOT / str(spec["data_inputs"][0])))
    resolved_font = resolve_font_stack(font_policy)
    configure_matplotlib(theme, font_policy, resolved_font)

    import matplotlib.pyplot as plt

    fig, axis = plt.subplots(
        1,
        1,
        figsize=(
            mm_to_inches(float(spec["size"]["width_mm"])),
            mm_to_inches(float(spec["size"]["height_mm"])),
        ),
        constrained_layout=bool(theme["layout"]["constrained"]),
    )
    apply_publication_layout(fig, theme)
    colors = {
        "Control": theme["palette"]["categorical"][0],
        "Treatment": theme["palette"]["categorical"][1],
    }
    markers = {"Batch1": "o", "Batch2": "s"}
    for row in rows:
        axis.scatter(
            row["pc1"],
            row["pc2"],
            s=42,
            color=colors[row["condition"]],
            marker=markers[row["batch"]],
            edgecolor=theme["palette"]["neutral"][0],
            linewidth=0.6,
        )
        if row["highlight_label"] == "yes":
            axis.text(
                row["pc1"] + 0.12,
                row["pc2"] + 0.12,
                row["sample_id"],
                fontsize=theme["typography"]["annotation_font_size_pt"],
                color=theme["palette"]["neutral"][0],
            )
    axis.text(
        float(theme["panel_labels"]["x_position_axes"]),
        float(theme["panel_labels"]["y_position_axes"]),
        "a",
        transform=axis.transAxes,
        fontsize=theme["panel_labels"]["font_size_pt"],
        fontweight=theme["panel_labels"]["font_weight"],
        va="top",
    )
    axis.axhline(0.0, color=theme["palette"]["neutral"][1], linewidth=0.8)
    axis.axvline(0.0, color=theme["palette"]["neutral"][1], linewidth=0.8)
    axis.set_title("Sample PCA", loc="left", pad=2)
    axis.set_xlabel("PC1 (48%)")
    axis.set_ylabel("PC2 (23%)")
    axis.spines["top"].set_visible(False)
    axis.spines["right"].set_visible(False)
    return fig


def build_figure(spec_path: Path) -> dict[str, str]:
    spec, theme, font_policy, profile, _ = project_resources(spec_path)
    rows = _load_rows(REPO_ROOT / str(spec["data_inputs"][0]))
    resolved_font = resolve_font_stack(font_policy)
    fig = create_figure(spec_path)
    output_dir = REPO_ROOT / str(spec["renderers"]["python"]["output_dir"])
    output_dir.mkdir(parents=True, exist_ok=True)
    stem = str(spec["figure_id"])
    svg_path = output_dir / f"{stem}.svg"
    pdf_path = output_dir / f"{stem}.pdf"
    png_path = output_dir / f"{stem}.png"
    fig.savefig(svg_path, format="svg")
    fig.savefig(pdf_path, format="pdf")
    fig.savefig(png_path, format="png", dpi=int(profile["preview_dpi"]))
    outputs = {
        "svg": str(svg_path.relative_to(REPO_ROOT)),
        "pdf": str(pdf_path.relative_to(REPO_ROOT)),
        "png": str(png_path.relative_to(REPO_ROOT)),
    }
    write_manifest(
        spec,
        profile,
        resolved_font,
        "python",
        spec_path,
        outputs,
        [
            "group_coloring_for_sample_separation",
            "batch_shape_encoding",
            "selective_sample_labels",
            "vector_first_export",
        ],
        {
            "panel_count": len(spec.get("panels", [])),
            "highlight_label_count": sum(
                1 for row in rows if str(row["highlight_label"]) == "yes"
            ),
            "reference_line_count": 2,
        },
    )
    return outputs


if __name__ == "__main__":
    raise SystemExit("Use figures/src/python/run_class_renderer.py")
